shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
   return FlutterAutofillTypeNone;
 }
 
+static BOOL isApproximatelyEqual(float x, float y, float delta) {
+  return x >= y - delta && x <= y + delta;
+}
+
+// Checks whether point should be considered closer to selectionRect compared to
+// otherSelectionRect.
+//
+// If checkRightBoundary is set, the right-center point on selectionRect and
+// otherSelectionRect will be used instead of the left-center point.
+//
+// This uses special (empirically determined) logic for determining the closer rect,
+// rather than a simple distance calculation. First, the closer vertical distance is
+// determined. Within the closest y distance, if the point is above the bottom
+// of the closest rect, the x distance will be minimized; however, if the point is
+// below the bottom of the rect, the x value will be maximized.
+static BOOL isSelectionRectCloserToPoint(CGPoint point,
+                                         CGRect selectionRect,
+                                         CGRect otherSelectionRect,
+                                         BOOL checkRightBoundary) {
+  CGPoint pointForSelectionRect =
+      CGPointMake(selectionRect.origin.x + (checkRightBoundary ? selectionRect.size.width : 0),
+                  selectionRect.origin.y + selectionRect.size.height * 0.5);
+  float yDist = fabs(pointForSelectionRect.y - point.y);
+  float xDist = fabs(pointForSelectionRect.x - point.x);
+
+  CGPoint pointForOtherSelectionRect =
+      CGPointMake(otherSelectionRect.origin.x + (checkRightBoundary ? selectionRect.size.width : 0),
+                  otherSelectionRect.origin.y + otherSelectionRect.size.height * 0.5);
+  float yDistOther = fabs(pointForOtherSelectionRect.y - point.y);
+  float xDistOther = fabs(pointForOtherSelectionRect.x - point.x);
+
+  BOOL isCloserVertically = yDist < yDistOther - 1;
```

Comment:
Why the `- 1` here?

Reply:
It's been awhile, but I think the idea was similar to `isApproximatelyEqual`, where there's a little bit of preference in favor of the `otherSelectionRect` to avoid unnecessary churn. I think where this sort of thing comes in most handy is when you have an emoji in the string, sometimes the center point of those is offset from the rest of the line a little bit.

Reply:
Maybe consider adding a comment explaining that.